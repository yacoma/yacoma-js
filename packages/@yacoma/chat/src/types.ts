import { AttachmentType } from '@yacoma/data'
import * as actions from './actions'
import * as effects from './effects'

export interface SelectedChat {
  // Selected collection id
  id: string
  // Force opening chat (do not toggle)
  force?: boolean
  // preview message
  preview?: boolean
}

export interface ChatState {
  draft?: {
    collectionId: string
    previewId?: string
    title: string
    attachments: {
      [attachIdx: string]: AttachmentType
    }
  }
}

export interface ChatConfig {
  state: {
    chat: ChatState
  }
  actions: {
    chat: typeof actions
  }
  effects: {
    chat: typeof effects
  }
}
