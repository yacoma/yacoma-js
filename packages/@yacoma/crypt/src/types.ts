// This **MUST** change if there are any breaking changes in API and the old
// version must be kept.
export const CRYPT_VERSION = 1

export interface EncryptedPayload {
  compress?: boolean
  buffer?: boolean
  iv: string
  cipher: string
}

export interface ClearPayload<T = any> {
  // extra fields passed as options
  payload: T
}
