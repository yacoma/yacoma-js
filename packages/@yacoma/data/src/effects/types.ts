import { UserWithKeys } from '@lucidogen/security'
import { GraphQLClient } from 'graphql-request'
import { SubscriptionClient } from 'subscriptions-transport-ws'
import { TransportSettings } from '../types'

export interface DataFxContext {
  user?: UserWithKeys
  client: GraphQLClient
  subscriptionClient: SubscriptionClient
  expire?: number
  sessionId?: string
  token: string
  endpoint: TransportSettings['endpoint']
  // TODO: helper functions
  // saveItem
  // etc
}
