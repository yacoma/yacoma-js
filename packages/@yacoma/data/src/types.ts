import { IContext } from 'overmind'
import { Base, User } from '@yacoma/security'
import * as actions from './actions'
import { Config, Context, Derive } from './app'
import {
  Collection,
  Contact,
  FileItem,
  Item,
  Message,
  PairCollection,
  WorkCollection,
} from './dataTypes'
import * as helpers from './helpers'

export interface ItemCallback<T> {
  (item: T): void
}

export interface AccessChanges {
  // The message to add when sharing to new people
  message?: string
  userAccess: {
    // false is set to remove userAccess entry
    [userId: string]: 'a' | 'e' | false
  }
  collectionAccess: {
    // True = keep or give access
    // False = remove access
    [collId: string]: boolean
  }
}

export interface BaseDataEffects {
  skipInit: boolean
  registerCallbacks(ctx: Context, app: IContext<Config>): void

  // Authentication
  // Registration disabled for now until we decide how it should
  // work for places where any user can create a login.
  register(ctx: Context): Promise<Contact | null>

  createContact(ctx: Context): Promise<Contact | null>

  login(ctx: Context): Promise<User | 'offline' | null>
  logout(ctx: Context): Promise<null>
  autoLogin(ctx: Context): Promise<User | 'offline' | null>

  // Query and subscribe
  query<T extends Base>(ctx: {}, type: string, variables?: any): Promise<T[]>
  subscribe(ctx: Context, type: string, variables?: any): void

  createPairCollection(ctx: Context, otherUser: User): Promise<Collection>
  createItem(
    ctx: Context,
    type: 'Item' | 'Collection' | 'File',
    content?: { [key: string]: any },
    opts?: { file?: File; collectionId?: string; open?: boolean }
  ): Promise<Item>
  createMessage(
    ctx: Context,
    collectionId: string,
    content?: any,
    preview?: boolean
  ): Promise<Message>
  saveItem<T extends Item = Item>(ctx: Context, theItem: T): Promise<T>

  inviteUsersToCollection(
    ctx: Context,
    sharedCollection: Collection,
    targetUsers: User[]
  ): Promise<Collection>
  shareItemToCollection(
    ctx: Context,
    sharedItem: Item,
    collectionId: string
  ): Promise<Item>

  saveAccessChanges(
    ctx: Context,
    item: Item,
    accessChanges: AccessChanges
  ): Promise<Item>
  download(ctx: Context, item: FileItem): Promise<void>
}

export type DataEffects = BaseDataEffects & typeof helpers

export interface SelectedDocument {
  id: string
  // Where we extracted the item from.
  collectionId: string
  // If true, do not toggle on reselect
  force?: boolean
}

export interface DataConfig {
  state: {
    document: {
      // Warning: There is more 'document' state defined in 'document' block
      search: string
      // how to view the collection content 'Messages', 'Items', 'Users', etc
      view: string
      sort?: 'title' | 'date'
      current: Derive<DataConfig['state']['document'], Item | undefined>
      selected?: SelectedDocument
      selectedCollectionId?: string
    }
    data: {
      login: {
        username: string
        email: string
        admin: boolean
        protected: boolean
        wordList: string[]
        wordListVerify: string[]
        error: string | undefined
        valid: Derive<DataConfig['state']['data']['login'], boolean>
        nameValid: Derive<DataConfig['state']['data']['login'], boolean>
      }
      token?: string
      // Always present. Set to empty values if not logged in.
      auth: User
      // Maximum number of bytes allowed for upload. Defaults to 4Mb
      uploadLimit: number
    }
    fileCache: {
      [fileId: string]: File
    }
    db: {
      Collection: {
        [collId: string]: Collection
      }
      PairCollection: {
        [collId: string]: PairCollection
      }
      WorkCollection: {
        [collId: string]: WorkCollection
      }
      PrivateCollection: {
        [collId: string]: Collection
      }
      Item: {
        [itemId: string]: Item
      }
      Contact: {
        [contactId: string]: Contact
      }
      User: {
        [userId: string]: User
      }
      File: {
        [itemId: string]: FileItem
      }
    }
    msgDb: {
      [collId: string]: {
        [messageId: string]: Message
      }
    }
  }
  actions: {
    data: typeof actions
  }
  effects: {
    data: DataEffects
  }
}

type GraphQL = string

export interface TypeHandler {
  // Data to load on app boot.
  query: GraphQL
  // GraphQL mutation to create, update, delete. A delete operation
  // is simply setting deleted field to true.
  mutation: GraphQL
  // GraphQL subscription to get updates on changed items.
  subscription: GraphQL
  // Sequence run when an Item is changed (created/updated/deleted or loaded from cache).
  callback?<T extends Base>(app: IContext<Config>): ItemCallback<T>
}

export interface AuthHandler {
  // Mutation to create a new contact (admin only):
  // {id, username, email, contact: string, privateKeys: string, publicKeys: string, privateCollection: string, workCollection: string }
  // Receives {contact: Contact} (not really needed: subscription will notify shared items)
  createContact: GraphQL

  // Mutation to create session
  // Receives {sessionId: string}
  // sessionId is used to create the Vasudeva token,
  // which is used in headers for all other queries
  mutateSession: GraphQL

  // Query to get the session from sessionId
  // Receives the full session with keys, userId, etc.
  querySession: GraphQL

  // Query to get the current user from userId.
  // Receives the full User with username, email, etc.
  authenticatedUser: GraphQL

  // Mutation to delete session
  logout: GraphQL
}

export interface TypeHandlers {
  // Declare types used in the app.
  [type: string]: TypeHandler
}

export interface TransportSettings {
  types: TypeHandlers
  endpoint: {
    // Just some string printed to console on connect.
    name?: string
    graphql: string
    cdn: string
    subscriptions: string
  }
  authentication: AuthHandler
  // When true, data does not run onInitialize.
  skipInit: boolean
  uploadLimit?: number
}

export interface DataSettings {
  data?: Partial<TransportSettings>
}

export interface SessionVariables {
  userId: string
  id: string
  type: 'Session'
  onlineAt: string
  keys: string
  token: string
  info?: string
}
