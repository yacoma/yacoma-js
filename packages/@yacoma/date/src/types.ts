import { Operator } from './app'

export interface DateConfig {
  state: {
    date: {
      // This value is updated every 15s. It is used to force a redraw of
      // components with date shown as `from now`.
      now: number
      // In [s]: how often `now` is updated. Defaults to 15 seconds.
      nowInterval: number
      // Starting day of the week for date picker or calendar
      // Defaults to (1 = Monday).
      weekDayStart: number
    }
  }

  actions: {
    date: {
      // This is run by a hook started on app boot.
      updateNow: Operator<void>
      // Change 'now' update interval (0 = stop updates).
      setInterval: Operator<number>
    }
  }

  effects: {
    date: {
      restart(interval: number): void
    }
  }
}
