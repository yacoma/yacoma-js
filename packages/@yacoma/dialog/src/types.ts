import { Operator } from './app'
import * as helpers from './helpers'

type DialogComponentType =
  | React.ComponentClass<{}>
  | React.FunctionComponent<{}>

export interface DialogComponentSettings {
  [key: string]: {
    component: DialogComponentType
    exclusive: boolean
  }
}

export interface DialogComponents {
  [key: string]: DialogComponentType
}

export interface DialogSettings {
  dialog?: {
    // Dialog components where we show the rest of the app.
    overlay?: DialogComponents
    // Dialog components where we hide the rest of the app.
    exclusive?: DialogComponents
  }
}

export interface DialogConfig {
  state: {
    dialog: {
      // Component to display as overlay dialog
      show?: string
      components: () => () => DialogComponentSettings
    }
  }
  actions: {
    dialog: {
      show: Operator<string>
      hide: Operator<void>
    }
  }
  effects: {
    dialog: typeof helpers
  }
}
