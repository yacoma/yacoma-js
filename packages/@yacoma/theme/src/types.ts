import { Derive } from './app'
import { defaultTheme } from './themes/defaultTheme'

type DefaultTheme = typeof defaultTheme

export interface Theme extends DefaultTheme {
  [key: string]: any
}

export interface ThemeConfig {
  state: {
    theme: {
      selected: string
      themes: Themes
      // Computed from selected, default and themes
      selectedTheme: Derive<Theme>
    }
  }
}

export interface Themes {
  [themeName: string]: {
    [Key: string]: any
  }
}

export interface ThemeSettings {
  theme?: Themes
}
