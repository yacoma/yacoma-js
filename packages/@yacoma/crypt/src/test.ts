import simple from 'simple-mock'

export type Describe = {
  (what: string, callback: () => void): void
  only: Describe
}

export interface It {
  (should: string, callback: () => void): void
  only: It
}

export interface Matcher {
  toBe(value: any): void
  toBeLessThanOrEqual(value: number): void
  toBeGreaterThanOrEqual(value: number): void
  toBeUndefined(): void
  toContain(value: any): void
  toEqual(value: any): void
  toMatchSnapshot(): void
  toMatch(re: RegExp): void
  toThrow(value: any): void
}

export type Expect = (value: any) => Matcher
export type FunctionCaller = (fn: () => void) => void
export type AfterAll = (fn: () => void) => void

declare var describe: Describe
declare var it: It
declare var expect: Expect
declare var beforeAll: FunctionCaller
declare var afterAll: FunctionCaller
declare var afterEach: FunctionCaller
declare var beforeEach: FunctionCaller

const afterAll_ = afterAll
const afterEach_ = afterEach
const beforeAll_ = beforeAll
const describe_ = describe
const expect_ = expect
const it_ = it
const beforeEach_ = beforeEach

export {
  afterAll_ as afterAll,
  afterEach_ as afterEach,
  beforeEach_ as beforeEach,
  beforeAll_ as beforeAll,
  describe_ as describe,
  expect_ as expect,
  it_ as it,
}

declare var global: any

export function mockRandomValues() {
  simple
    .mock(global.crypto, 'getRandomValues')
    .callFn(function(bytes: Uint8Array) {
      for (let i = 0; i < bytes.length; ++i) {
        bytes[i] = i
      }
    })
}

export function restore() {
  simple.restore()
}
