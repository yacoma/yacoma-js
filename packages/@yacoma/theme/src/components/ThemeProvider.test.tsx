import * as React from 'react'
import { describe, expect, it } from 'test'

import { build, settings } from '@lucidogen/build'

import { theme } from '../'
import { defaultTheme } from '../themes'
import { ThemeSettings } from '../types'
import {
  DirectThemeProvider as directThemeProvider,
  ThemeProvider as themeProvider,
} from './ThemeProvider'

const main = {
  name: 'main',
  settings: settings<ThemeSettings>({
    theme: {
      default: { thisWidth: '12px' },
    },
  }),
}

const app = build(main)
  .using(theme)
  .app()

describe('ThemeProvider', () => {
  it('should render children', () => {
    const connect = createConnect(app)
    const ThemeProvider = connect(themeProvider)
    // TODO: Should we mock `injectGlobal` from styled-components ?
    expect(
      testRender(<ThemeProvider scopeName="foo">foobar</ThemeProvider>)
    ).toMatchSnapshot()
  })
})

describe('DirectThemeProvider', () => {
  it('should render with provided theme', () => {
    const connect = createConnect(app)
    const DirectThemeProvider = connect(directThemeProvider)
    expect(
      testRender(
        <DirectThemeProvider scopeName="bar" theme={defaultTheme}>
          foobar
        </DirectThemeProvider>
      )
    ).toMatchSnapshot()
  })

  /*
  it('should render without connected', () => {
    const DirectThemeProvider = directThemeProvider
    expect(
      testRender(
        <DirectThemeProvider scopeName="bar" theme={defaultTheme}>
          foobar
        </DirectThemeProvider>
      )
    ).toMatchSnapshot()
  })
  */
})
