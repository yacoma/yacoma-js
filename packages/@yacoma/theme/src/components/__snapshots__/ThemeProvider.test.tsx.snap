// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectThemeProvider should render with provided theme 1`] = `
<div
  className="theme-bar"
>
  foobar
</div>
`;

exports[`DirectThemeProvider should render without connected 1`] = `
<div
  className="theme-bar"
>
  foobar
</div>
`;

exports[`ThemeProvider should render children 1`] = `
<div
  className="theme-foo"
>
  foobar
</div>
`;
